src/lib/include/simd/simd_float_datastructures.hpp

```diff
@@ -0,0 +1,178 @@
+#ifndef DISTOPIA_SIMD_FLOAT_DATASTRUCTURES_H
+#define DISTOPIA_SIMD_FLOAT_DATASTRUCTURES_H
+
+#include "simd_config.h"
+#include "simd_doubles.h"
+#include "simd_floats.h"
+#include <immintrin.h>
+
+// 3x8 structure of packed floats
+class SimdFloatX8X3 {
+public:
+  SimdFloatX8X3(SimdFloatX8 a, SimdFloatX8 b, SimdFloatX8 c);
+  inline void JNG_AOS2SOASwizzle();
+  inline void AOS2SOASwizzle();
+  inline void SOA2AOSSwizzle();
+  SimdFloatX8 x;
+  SimdFloatX8 y;
+  SimdFloatX8 z;
+};
+
+SimdFloatX8X3::SimdFloatX8X3(SimdFloatX8 a, SimdFloatX8 b, SimdFloatX8 c)
+    : x(a), y(b), z(c) {}
+
+/* performs an in place transpose of the data in a, b and c from
+SOA to AOS form.   x,y,z,x,y,z.. to  x0-7,y0-7,z0-7  */
+inline void SimdFloatX8X3::AOS2SOASwizzle() {
+  /* based on Intel 3D vector norm with intrinsics */
+
+  /* x = x0y0z0x1y1z1x2y2 */
+  /* y = z2x3y3z3x4y4z4x5 */
+  /* z = y6z6x7y7z7x8y8z8 */
+
+  /* break into smaller chunks to swizzle, this may be slower than JNG
+    implementation */
+  __m128 m0 = _mm256_extractf128_ps(x.contents, 0);
```

Comment:
Avoid `_mm256_extractf128_ps(foo, 0)` to extract the low half of the vector: it has high-latency and can only run on one port. [`_mm256_castps256_ps128`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#techs=AVX&text=_mm256_castps256_ps128&expand=5189,5189,622) does the same thing but it’s free: it does not compile to any instructions because of how the AVX registers work.

Reply:
awesome, but must be used for high part right?

Reply:
Yep, that’s correct. The reason why is that 256-bit (`ymm`) registers are shared with 128-bit (`xmm`) registers.

For example you can set `ymm0 = {0.f, 1.f, 2.f, 3.f, 4.f, 5.f, 6.f, 7.f}`. Then the instruction
```asm
vsqrtps %ymm0, %ymm1
```
calculates the square root of every element of `ymm0` and places it in `ymm1`. But if instead you ran
```asm
vsqrtps %xmm0, %xmm1
```
your calculation would still succeed, although it would only process the low half of the vector! `xmm0` is simply an alias for the low half of `ymm0`, so by writing to `ymm0` we’re writing to `xmm0` as well.

The `_mm256_castps256_ps128` function is free because the compiler can just write `xmm` whenever it wants to use the low half of a `ymm` register.